File: ai/src/test/java/org/conductoross/conductor/ai/LLMsTest.java
package org.conductoross.conductor.ai;

import java.util.List;

import org.conductoross.conductor.ai.models.EmbeddingGenRequest;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import com.netflix.conductor.common.metadata.tasks.Task;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.*;

class LLMsTest {

    private AIModelProvider mockModelProvider;
    private LLMs llms;
    private AIModel mockModel;

    @BeforeEach
    void setUp() {
        mockModelProvider = mock(AIModelProvider.class);
        mockModel = mock(AIModel.class);
        when(mockModelProvider.getPayloadStoreLocation()).thenReturn("/tmp/test-payload");
        when(mockModelProvider.getModel(any())).thenReturn(mockModel);

        llms = new LLMs(List.of(), null, mockModelProvider, new okhttp3.OkHttpClient());
    }

    @Test
    void testConstructor_setsPayloadStoreLocation() {
        assertEquals("/tmp/test-payload", llms.payloadStoreLocation);
    }

    @Test
    void testGenerateEmbeddings_delegatesToModel() {
        Task mockTask = mock(Task.class);
        when(mockTask.getWorkflowInstanceId()).thenReturn("wf-1");
        when(mockTask.getTaskId()).thenReturn("task-1");

        List<Float> expectedEmbeddings = List.of(0.1f, 0.2f, 0.3f);
        when(mockModel.generateEmbeddings(any())).thenReturn(expectedEmbeddings);

        EmbeddingGenRequest request = new EmbeddingGenRequest();
        request.setLlmProvider("openai");

        List<Float> result = llms.generateEmbeddings(mockTask, request);

        assertEquals(expectedEmbeddings, result);
        verify(mockModel).generateEmbeddings(request);
    }

    @Test
    void testGenerateEmbeddings_usesCorrectProvider() {
        Task mockTask = mock(Task.class);
        when(mockTask.getWorkflowInstanceId()).thenReturn("wf-1");
        when(mockTask.getTaskId()).thenReturn("task-1");

        when(mockModel.generateEmbeddings(any())).thenReturn(List.of(0.5f));

        EmbeddingGenRequest request = new EmbeddingGenRequest();
        request.setLlmProvider("anthropic");
        request.setModel("text-embedding-model");
        request.setText("Sample text for embedding");

        llms.generateEmbeddings(mockTask, request);

        verify(mockModelProvider).getModel(request);
    }
}

File: ai/src/test/typescript/LLMs.test.ts
import { describe, it, expect, vi, beforeEach } from 'vitest';
import { LLMs } from '../../main/typescript/LLMs.js';
import { AIModelProvider } from '../../main/typescript/AIModelProvider.js';
import { AIModel } from '../../main/typescript/AIModel.js';
import { EmbeddingGenRequest } from '../../main/typescript/models/index.js';
import { Task } from '../../main/typescript/types/index.js';

describe('LLMs', () => {
  let mockModelProvider: any;
  let mockModel: any;
  let llms: any;

  beforeEach(() => {
    mockModel = {
      generateEmbeddings: vi.fn(),
    };

    mockModelProvider = {
      payloadStoreLocation: '/tmp/test-payload',
      getModel: vi.fn().mockReturnValue(mockModel),
    };

    llms = new LLMs([], {} as any, mockModelProvider as unknown as AIModelProvider);
  });

  it('testConstructor_setsPayloadStoreLocation', () => {
    expect((llms as any).payloadStoreLocation).toBe('/tmp/test-payload');
  });

  it('testGenerateEmbeddings_delegatesToModel', async () => {
    const mockTask: Task = {
      workflowInstanceId: 'wf-1',
      taskId: 'task-1',
      taskDefName: 'test-task',
      taskType: 'test',
      status: 'IN_PROGRESS',
    };

    const expectedEmbeddings = [0.1, 0.2, 0.3];
    mockModel.generateEmbeddings.mockResolvedValue(expectedEmbeddings);

    const request: EmbeddingGenRequest = {
      llmProvider: 'openai',
    };

    const result = await llms.generateEmbeddings(mockTask, request);

    expect(result).toEqual(expectedEmbeddings);
    expect(mockModel.generateEmbeddings).toHaveBeenCalledWith(request);
  });

  it('testGenerateEmbeddings_usesCorrectProvider', async () => {
    const mockTask: Task = {
      workflowInstanceId: 'wf-1',
      taskId: 'task-1',
      taskDefName: 'test-task',
      taskType: 'test',
      status: 'IN_PROGRESS',
    };

    mockModel.generateEmbeddings.mockResolvedValue([0.5]);

    const request: EmbeddingGenRequest = {
      llmProvider: 'anthropic',
      model: 'text-embedding-model',
      text: 'Sample text for embedding',
    };

    await llms.generateEmbeddings(mockTask, request);

    expect(mockModelProvider.getModel).toHaveBeenCalledWith(request);
  });
});
